File: _legacy-swift/Models/AppError.swift
import Foundation

/// Errors that surface to the user. Each case maps to a localized message key.
public enum AppError: LocalizedError, Equatable, Sendable {
    case photoPermissionDenied
    case videoUnsupported
    case videoTooLong(maxSeconds: Int)
    case exportFailed(underlying: String)
    case adUnavailable
    case purchaseFailed(underlying: String)

    public var errorDescription: String? {
        switch self {
        case .photoPermissionDenied:
            return NSLocalizedString("error.photo_permission", comment: "")
        case .videoUnsupported:
            return NSLocalizedString("error.video_unsupported", comment: "")
        case .videoTooLong:
            return NSLocalizedString("error.video_too_long", comment: "")
        case .exportFailed:
            return NSLocalizedString("error.export_failed", comment: "")
        case .adUnavailable:
            return NSLocalizedString("error.ad_unavailable", comment: "")
        case .purchaseFailed:
            return NSLocalizedString("error.purchase_failed", comment: "")
        }
    }
}

File: src/types/AppError.ts
/**
 * Discriminated union of user-facing errors. Each kind maps to a localized
 * message key in `src/i18n/*`.
 */
export type AppError =
  | { kind: 'photoPermissionDenied' }
  | { kind: 'videoUnsupported' }
  | { kind: 'videoTooLong'; maxSeconds: number }
  | { kind: 'exportFailed'; underlying: string }
  | { kind: 'adUnavailable' }
  | { kind: 'purchaseFailed'; underlying: string }
  | { kind: 'nativeModuleUnavailable' };

export const appErrorMessageKey = (error: AppError): string => {
  switch (error.kind) {
    case 'photoPermissionDenied':
      return 'error.photo_permission';
    case 'videoUnsupported':
      return 'error.video_unsupported';
    case 'videoTooLong':
      return 'error.video_too_long';
    case 'exportFailed':
      return 'error.export_failed';
    case 'adUnavailable':
      return 'error.ad_unavailable';
    case 'purchaseFailed':
      return 'error.purchase_failed';
    case 'nativeModuleUnavailable':
      return 'error.native_unavailable';
  }
};
